### ISSUE TITLE
Need for clarification

### ISSUE THREAD

**Issue by yota-code** [webmproject/codec-compare#3]
Hello,

I must say your representation is a bit hard to follow for a novice :D

When you look at this page : https://storage.googleapis.com/demos.webmproject.org/webp/cmp/2024_01_25/index.html?load=jpegxl_vs_avif.json#ref=AVIF+speed+6

avif seems to have the lead over jpegxl at every efforts

When you look at this one : https://storage.googleapis.com/demos.webmproject.org/webp/cmp/2024_01_25/index.html?load=all.json#WebP+method+4-quality=95..95&matcher_ssim=off

The tendency is flipped, with jpegxl taking the lead ?

Could you explain how to use your site to have a good idea on the relative performance of both codec ?

Also, it is difficult to get an idea of the quality selected as reference, perhaps a link to an image at this quality could help showing the level of distortion obtained (I guess < 0.1bpp can not seriously be used for display)

**Comment by y-guyon** [comment]
Thank you for the feedback!

> I must say your representation is a bit hard to follow for a novice :D

I agree and I heard that multiple times already. I will try to improve that side of the framework at some point.

> When you look at this page : https://storage.googleapis.com/demos.webmproject.org/webp/cmp/2024_01_25/index.html?load=jpegxl_vs_avif.json#ref=AVIF+speed+6
> 
> avif seems to have the lead over jpegxl at every efforts
> 
> When you look at this one : https://storage.googleapis.com/demos.webmproject.org/webp/cmp/2024_01_25/index.html?load=all.json#WebP+method+4-quality=95..95&matcher_ssim=off
> 
> The tendency is flipped, with jpegxl taking the lead ?

You are correct. As for any data visualization tool, the results can greatly vary depending on the constraints applied to the comparison. One can easily skew the findings in one or another way, intentionally or inadvertently.

I noticed that your second plot only has 2521 comparisons, which may signal low confidence in the results due to the lack of data points.

> Could you explain how to use your site to have a good idea on the relative performance of both codec ?

Sure, if you specify _relative to what_.  
The goal of this framework is not to prove that "codec X is better". The purpose is showing that "codec X is A% faster to encode, B% slower to decode, generates files C% smaller than codec Y, for the images in this data set with the same metric value, on this quality range etc."  
Having stronger corpus or environment assumptions could be misleading.

I tried to minimize any bias in the [default settings](https://storage.googleapis.com/demos.webmproject.org/webp/cmp/2024_01_25/index.html):
- Two modern codecs AVIF and JPEG XL, both compared to WebP which has been stable in both format and implementation for years now.
- The effort/speed/method encoding settings (meaning how much time/CPU the user would like to spend for encoding an image) is the same as the default in the binaries of the main codec implementations (cwebp, avifenc, cjxl).
- The full input quality setting ranges are used.
- Comparing bytes and timings is easier to reason with: "twice faster" and "20% smaller" are more accessible than "better looking by 2dB". That is why comparisons are made for the same visual quality, according to some objective metric.
- There is no consensus on the objective metric to rely on. [SSIM favors AVIF](https://storage.googleapis.com/demos.webmproject.org/webp/cmp/2024_01_25/index.html?load=jpegxl_vs_avif.json#ref=AVIF+speed+6&matcher_ssimulacra2=off) and [SSIMULACRA2 favors JPEG XL](https://storage.googleapis.com/demos.webmproject.org/webp/cmp/2024_01_25/index.html?load=jpegxl_vs_avif.json#ref=AVIF+speed+6&matcher_ssim=off). To avoid picking a side, I enabled both by default. Unfortunately this approach reduces the total number of comparisons and probably skews the plot in its own way, but this is somewhat compensated by the fact that the interactive tool allows for easy tuning of matchers and metrics to fit one's needs or convictions.
- Subjective visual quality evaluation (aka user surveys) could settle the debate for a given corpus and settings. Unfortunately this is rather involved.

> Also, it is difficult to get an idea of the quality selected as reference, perhaps a link to an image at this quality could help showing the level of distortion obtained

This heavily depends on the image unfortunately. I agree that showing one image at a given quality setting or bpp would give a sense of the subjective visual quality loss, but it would also be misleading for other images that may require a different amount of quantization to achieve the same "loss" (if such a definition exists).

In codec-compare one can see the images by clicking on a small dot, scrolling down and clicking on the image below. I was thinking about making this more visible.

> I guess < 0.1bpp can not seriously be used for display

From the [<0.1bpp plot](https://storage.googleapis.com/demos.webmproject.org/webp/cmp/2024_01_25/index.html?batch=batches/avif_6.json&batch=batches/jpegxl_7.json#AVIF+speed+6-bpp=0..0.1):  
[Top left point](https://storage.googleapis.com/demos.webmproject.org/webp/cmp/2024_01_25/visualizer.html?bimg=..%2Fclic_validation_2021_2022_2024%2Fimages%2Fcf79510012b96c7326c34eba548ea02b1e47097402fc872465679dd9e6dbc4ac.png&btxt=original&rimg=encoded%2Fcf79510012b96c7326c34eba548ea02b1e47097402fc872465679dd9e6dbc4ac.e6q025.avif&rtxt=AVIF+speed+6&limg=encoded%2Fcf79510012b96c7326c34eba548ea02b1e47097402fc872465679dd9e6dbc4ac.e7q000.jxl.png&ltxt=JPEG+XL+effort+7): bpp < 0.03  
[Right-most point](https://storage.googleapis.com/demos.webmproject.org/webp/cmp/2024_01_25/visualizer.html?bimg=..%2Fclic_validation_2021_2022_2024%2Fimages%2F9e248324151508f3e4790532329acca967cb9e0484f4fea697bab3a3957debd8.png&btxt=original&rimg=encoded%2F9e248324151508f3e4790532329acca967cb9e0484f4fea697bab3a3957debd8.e6q017.avif&rtxt=AVIF+speed+6&limg=encoded%2F9e248324151508f3e4790532329acca967cb9e0484f4fea697bab3a3957debd8.e7q002.jxl.png&ltxt=JPEG+XL+effort+7): bpp < 0.03  
Not super crisp, but can be displayed as a background for example.

**Comment by y-guyon** [comment]
> Which SSIM code did you use ?

It is written here:

![image](https://github.com/webmproject/codec-compare/assets/51164219/074123f6-8fa8-4046-8eb7-67af336f98d2)

> **libwebp2 quality metrics command**  
> Command line to get libwebp2 quality metrics PSNR and SSIM:
> ```sh
> git clone https://chromium.googlesource.com/codecs/libwebp2
> cd libwebp2
> git checkout e0c6533
> cd ..
> cmake -S libwebp2 -B libwebp2/build -DCMAKE_C_COMPILER=clang -DCMAKE_CXX_COMPILER=clang++
> cmake --build libwebp2/build --parallel
> libwebp2/build/get_disto -ssim ${encoded_name}.png ${original_name}
> ```

So it would be https://chromium.googlesource.com/codecs/libwebp2/+/e0c6533107649063c236b5666f2cc4c10b9b7591/src/enc/distortion_enc.cc#418

**Comment by y-guyon** [comment]
> I agree there is no consensus about the metric, yet ssimulacra2 is the only one which was fitted on subjective data (which does not make it perfect everywhere but still)

The issue with picking any single metric by default is that it is unfair to all codecs that do not optimize for that in their Rate-Distortion optimization.

libavif uses aom's SSIM [by default](https://github.com/AOMediaCodec/libavif/blob/b10d2697e9ed2fb09cb722335ff4342c353612b8/src/codec_aom.c#L876) but allows for PSNR to be used.  
libwebp uses PSNR for RD-opt.  
Do you know if libjxl supports tuning the metric used for RD-opt? It uses SSIMULACRA2.1 by default right?

**Comment by yota-code** [comment]
I'm not sure but I believe it uses the [butteraugli](https://github.com/google/butteraugli) metric as an internal target (hence a clear focus on high quality compression).
If you want to know more I invite you to jump on the discord :) ( link here: https://jpegxl.info/ )

Many core developers are there, even some from google (Zürich office) and even one which worked on the webp lossless codec :) and they all enjoy technical discussions

**Comment by y-guyon** [comment]
The [2024_03_25 batch](https://storage.googleapis.com/demos.webmproject.org/webp/cmp/2024_03_25/index.html) uses SSIM and Butteraugli as discussed above. Feel free to reopen this thread if necessary.
